File: sdk-libs/transaction/src/wallet/state.rs
use std::{
    cmp::Ordering,
    collections::{hash_map::Entry, BTreeSet, HashMap, HashSet},
};

use solana_address::Address;
use solana_signature::Signature;
use zolana_keypair::{shielded::ShieldedAddress, viewing_key::ViewTag, P256Pubkey};

use crate::{
    error::TransactionError, instructions::transact::OutputContext, utxo::Utxo, AssetRegistry,
};

pub const DEFAULT_TAG_WINDOW: u64 = 64;
pub(crate) const SENDER_HISTORY_ROW_BASE: u64 = 1 << 63;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct PrivateTransactionId {
    pub signature: String,
    pub slot: u64,
    /// Stable row discriminator within the transaction. For received outputs this
    /// is the UTXO leaf index when available; sender-side aggregate rows use a
    /// high local row index range.
    pub index: u64,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum PrivateTransactionKind {
    Deposit,
    PrivateTransfer,
    PublicWithdrawal,
    Split,
    Merge,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum PrivateTransactionDirection {
    Inbound,
    Outbound,
    SelfTransfer,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum PrivateTransactionStatus {
    Confirmed,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct PrivateTransaction {
    pub id: PrivateTransactionId,
    pub kind: PrivateTransactionKind,
    pub direction: PrivateTransactionDirection,
    pub status: PrivateTransactionStatus,
    pub asset: Address,
    pub amount: u64,
    pub counterparty_viewing_pubkey: Option<P256Pubkey>,
}

pub struct ViewingKeyEntry {
    pub viewing_pubkey: P256Pubkey,
    pub created_at: i64,
    pub tx_count: u64,
    pub request_count: u64,
    pub known_senders: HashMap<P256Pubkey, u64>,
    pub known_recipients: HashMap<P256Pubkey, u64>,
}

impl ViewingKeyEntry {
    pub fn new(viewing_pubkey: P256Pubkey, created_at: i64) -> Self {
        Self {
            viewing_pubkey,
            created_at,
            tx_count: 0,
            request_count: 0,
            known_senders: HashMap::new(),
            known_recipients: HashMap::new(),
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct WalletUtxo {
    pub utxo: Utxo,
    pub output_context: OutputContext,
    pub nullifier: [u8; 32],
    pub data_hash: Option<[u8; 32]>,
    pub ring_data_hash: Option<[u8; 32]>,
    pub spent: bool,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct AssetBalance {
    pub asset_id: u64,
    pub mint: Address,
    pub amount: u64,
    pub utxos: Vec<Utxo>,
}

/// Spendable default-ring balances.
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct Balances {
    pub assets: Vec<AssetBalance>,
}

/// Holdings bound to one ring, never selectable by the default spend path.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct RingBalance {
    pub ring_program_id: Address,
    pub assets: Vec<AssetBalance>,
}

impl Balances {
    pub fn get_balance(&self, mint: Address) -> Option<&AssetBalance> {
        self.assets.iter().find(|balance| balance.mint == mint)
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Filter {
    MinAmount(u64),
}

impl Filter {
    fn matches(&self, utxo: &Utxo) -> bool {
        match self {
            Filter::MinAmount(min) => utxo.amount >= *min,
        }
    }
}

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct SyncReport {
    pub stored_utxos: usize,
    pub unparsed_transactions: usize,
    pub undecryptable_candidates: usize,
    /// Compact asset ids that failed to decode because the wallet's registry
    /// did not know them (SPL assets registered after the registry was built).
    /// The client sync layer uses this to lazily backfill the registry from
    /// chain and retry; it stays empty when every id is known.
    pub unknown_asset_ids: BTreeSet<u64>,
}

/// One transaction's place in the total order every rings stream shares.
///
/// Ordered by `(slot, signature)`, matching the indexer's pagination. Record a
/// position only once every row of that transaction has been applied, a resume
/// returns rows strictly after it.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub struct ChainPosition {
    pub slot: u64,
    pub signature: Signature,
}

impl Ord for ChainPosition {
    fn cmp(&self, other: &Self) -> Ordering {
        (self.slot, self.signature.as_ref()).cmp(&(other.slot, other.signature.as_ref()))
    }
}

impl PartialOrd for ChainPosition {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

/// One key on one indexer stream. `ViewTag` is `[u8; 32]`, so the variant is
/// what stops a tag being read as a nullifier.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum CursorStream {
    /// Shielded transactions matched by output view tag.
    Tags(ViewTag),
    /// Shielded transactions matched by spent nullifier.
    Nullifiers([u8; 32]),
    /// Encrypted UTXOs, which proofless deposits are read from.
    Proofless(ViewTag),
}

impl CursorStream {
    /// The tag or nullifier, for building the query.
    pub fn value(self) -> [u8; 32] {
        match self {
            Self::Tags(value) | Self::Nullifiers(value) | Self::Proofless(value) => value,
        }
    }
}

pub struct Wallet {
    /// Public wallet identity. All secret key material is supplied by a
    /// `WalletAuthority` when cryptographic work is required.
    pub identity: ShieldedAddress,
    /// Asset-id ↔ mint translation config for this wallet's session. Built once
    /// before the wallet and immutable afterward; the build and sync paths read
    /// it to encode/decode UTXO asset ids.
    pub registry: AssetRegistry,
    pub viewing_key_history: Vec<ViewingKeyEntry>,
    pub utxos: Vec<WalletUtxo>,
    pub transactions: Vec<PrivateTransaction>,
    /// Every input nullifier ever observed across synced transactions. Kept
    /// permanently so a UTXO discovered after its spend was seen still marks
    /// spent.
    pub nullifiers: HashSet<[u8; 32]>,
    pub last_synced: i64,
    /// Per key, the position everything matching it has been seen through.
    /// Streams advance independently. Nullifier entries die with their spend.
    pub cursors: HashMap<CursorStream, ChainPosition>,
}

impl Wallet {
    pub fn new(
        identity: ShieldedAddress,
        registry: AssetRegistry,
    ) -> Result<Self, TransactionError> {
        let viewing_pubkey = identity.viewing_pubkey;
        Ok(Self {
            identity,
            registry,
            viewing_key_history: vec![ViewingKeyEntry::new(viewing_pubkey, 0)],
            utxos: Vec::new(),
            transactions: Vec::new(),
            nullifiers: HashSet::new(),
            last_synced: 0,
            cursors: HashMap::new(),
        })
    }

    pub(crate) fn ensure_viewing_key_entries(
        &mut self,
        viewing_pubkeys: impl IntoIterator<Item = P256Pubkey>,
    ) {
        for viewing_pubkey in viewing_pubkeys {
            if self
                .viewing_key_history
                .iter()
                .all(|entry| entry.viewing_pubkey != viewing_pubkey)
            {
                self.viewing_key_history
                    .push(ViewingKeyEntry::new(viewing_pubkey, 0));
            }
        }
    }

    /// Every viewing key this wallet has been given, current and rotated-out.
    ///
    /// Seeded from the identity and extended by [`Self::ensure_viewing_key_entries`]
    /// on each sync, so it also holds keys a later scan's material omits. A scan
    /// snapshots this before it borrows the wallet mutably; a transfer addressed
    /// to any of these keys is addressed to this wallet.
    pub(crate) fn self_viewing_pubkeys(&self) -> HashSet<P256Pubkey> {
        self.viewing_key_history
            .iter()
            .map(|entry| entry.viewing_pubkey)
            .collect()
    }

    pub fn private_transactions(&self) -> &[PrivateTransaction] {
        &self.transactions
    }

    pub fn get_private_transactions(&self) -> Vec<PrivateTransaction> {
        self.transactions.clone()
    }

    pub(super) fn unspent(&self) -> impl Iterator<Item = &WalletUtxo> {
        self.utxos.iter().filter(|u| !u.spent)
    }

    /// The spendable default-ring balance of one mint.
    pub fn balance(
        &self,
        mint: Address,
        filter: Option<Filter>,
    ) -> Result<AssetBalance, TransactionError> {
        let mut balance = AssetBalance {
            asset_id: self.registry.asset_id(&mint)?,
            mint,
            amount: 0,
            utxos: Vec::new(),
        };
        for wallet_utxo in self.unspent() {
            if wallet_utxo.utxo.asset != mint || wallet_utxo.utxo.ring_program_id.is_some() {
                continue;
            }
            if let Some(filter) = &filter {
                if !filter.matches(&wallet_utxo.utxo) {
                    continue;
                }
            }
            balance.amount = balance.amount.saturating_add(wallet_utxo.utxo.amount);
            balance.utxos.push(wallet_utxo.utxo.clone());
        }
        Ok(balance)
    }

    /// Spendable default-ring balances, ring-bound notes appear in
    /// [`Self::ring_balances`].
    pub fn balances(&self, skip_utxos: bool) -> Result<Vec<AssetBalance>, TransactionError> {
        self.asset_balances(skip_utxos, |entry| entry.utxo.ring_program_id.is_none())
    }

    pub fn ring_balances(&self, skip_utxos: bool) -> Result<Vec<RingBalance>, TransactionError> {
        let rings: BTreeSet<Address> = self
            .unspent()
            .filter_map(|entry| entry.utxo.ring_program_id)
            .collect();
        rings
            .into_iter()
            .map(|ring| {
                Ok(RingBalance {
                    ring_program_id: ring,
                    assets: self.asset_balances(skip_utxos, |entry| {
                        entry.utxo.ring_program_id == Some(ring)
                    })?,
                })
            })
            .collect()
    }

    fn asset_balances(
        &self,
        skip_utxos: bool,
        eligible: impl Fn(&WalletUtxo) -> bool,
    ) -> Result<Vec<AssetBalance>, TransactionError> {
        let mut by_mint: HashMap<Address, AssetBalance> = HashMap::new();
        for wallet_utxo in self.unspent() {
            if !eligible(wallet_utxo) {
                continue;
            }
            let balance = match by_mint.entry(wallet_utxo.utxo.asset) {
                Entry::Occupied(occupied) => occupied.into_mut(),
                Entry::Vacant(vacant) => vacant.insert(AssetBalance {
                    asset_id: self.registry.asset_id(&wallet_utxo.utxo.asset)?,
                    mint: wallet_utxo.utxo.asset,
                    amount: 0,
                    utxos: Vec::new(),
                }),
            };
            balance.amount = balance.amount.saturating_add(wallet_utxo.utxo.amount);
            if !skip_utxos {
                balance.utxos.push(wallet_utxo.utxo.clone());
            }
        }
        let mut balances: Vec<AssetBalance> = by_mint.into_values().collect();
        balances.sort_by_key(|b| b.asset_id);
        Ok(balances)
    }
}

File: sdk-libs/ts/src/transaction/wallet/state.ts
import type { Address, Bytes32, Signature } from "../../interface/types.js";
import type { P256PublicKey } from "../../keypair/public-key.js";
import type { ShieldedAddress } from "../../keypair/shielded.js";

import { TransactionError } from "../error.js";
import { copy } from "../internal.js";
import { Utxo } from "../utxo.js";
import { AssetRegistry } from "./asset.js";

export interface AssetBalance {
  readonly assetId: bigint;
  readonly mint: Address;
  readonly amount: bigint;
  readonly utxos: readonly Utxo[];
}

/** Narrows which unspent notes a balance counts. */
export type Filter = Readonly<{ kind: "minAmount"; minAmount: bigint }>;

function matches(filter: Filter, utxo: Utxo): boolean {
  return utxo.amount >= filter.minAmount;
}

/**
 * Stable identity of one history row. `index` discriminates rows within a
 * transaction: received outputs use the UTXO leaf index where the indexer
 * supplies one, and sender-side aggregate rows use a high local range.
 */
export interface PrivateTransactionId {
  readonly signature: Signature;
  readonly slot: bigint;
  readonly index: bigint;
}

/**
 * Sender-side aggregate rows are indexed from here, above every leaf index a
 * tree can hand out, so they cannot collide with a received row.
 */
export const SENDER_HISTORY_ROW_BASE = 1n << 63n;

export type PrivateTransactionKind =
  | "deposit"
  | "privateTransfer"
  | "publicWithdrawal"
  | "split"
  | "merge";

export type PrivateTransactionDirection = "inbound" | "outbound" | "selfTransfer";

/**
 * Streams the wallet keeps read positions in.
 *
 * They are separate because reaching the tip of one says nothing about the
 * others, and sharing a cursor would skip rows in whichever stream is behind.
 * `transactions` and `proofless` are keyed by view tag; `nullifiers` by
 * nullifier.
 */
export type CursorStream = "transactions" | "proofless" | "nullifiers";

/**
 * A history row is reconstructed from an indexed transaction, so it exists only
 * once that transaction has landed. Nothing stages a locally submitted transfer
 * into the history ahead of a sync, in either language, so `confirmed` is the
 * only state a row can be in.
 */
export type PrivateTransactionStatus = "confirmed";

export interface PrivateTransaction {
  readonly id: PrivateTransactionId;
  readonly kind: PrivateTransactionKind;
  readonly direction: PrivateTransactionDirection;
  readonly status: PrivateTransactionStatus;
  readonly asset: Address;
  readonly amount: bigint;
  readonly counterpartyViewingPublicKey?: P256PublicKey;
}

export interface SyncReport {
  readonly storedUtxos: number;
  readonly unparsedTransactions: number;
  readonly undecryptableCandidates: number;
  /**
   * Compact asset ids that failed to decode because the wallet's registry did
   * not know them, ascending. The client sync layer uses this to lazily
   * backfill the registry from chain and retry; it stays empty when every id is
   * known.
   */
  readonly unknownAssetIds: readonly bigint[];
  /**
   * Merge asset fields that could not be resolved through the wallet's
   * registry. The client sync layer backfills the registry only when this or
   * `unknownAssetIds` is non-empty.
   */
  readonly unknownAssetFields: readonly Bytes32[];
}

/**
 * A viewing public key retained for historical deposit discovery and
 * decryption after key rotation.
 */
export interface ViewingKeyEntry {
  readonly viewingPublicKey: P256PublicKey;
  readonly createdAt: bigint;
}

export function newViewingKeyEntry(
  viewingPublicKey: P256PublicKey,
  createdAt: bigint,
): ViewingKeyEntry {
  return Object.freeze({
    viewingPublicKey,
    createdAt,
  });
}

function snapshotViewingKeyEntry(value: ViewingKeyEntry): ViewingKeyEntry {
  return Object.freeze({
    viewingPublicKey: value.viewingPublicKey,
    createdAt: value.createdAt,
  });
}

export interface WalletUtxo {
  readonly utxo: Utxo;
  readonly outputContext: Readonly<{
    hash: Bytes32;
    tree: Address;
    leafIndex: bigint;
  }>;
  readonly nullifier: Bytes32;
  readonly dataHash?: Bytes32;
  readonly zoneDataHash?: Bytes32;
  readonly spent: boolean;
}

function copyUtxo(value: Utxo): Utxo {
  return new Utxo({
    owner: value.owner,
    asset: value.asset,
    amount: value.amount,
    blinding: value.blinding,
    data: value.data,
    ...(value.zoneProgramId === undefined ? {} : { zoneProgramId: value.zoneProgramId }),
  });
}

function snapshotUtxo(value: WalletUtxo): WalletUtxo {
  return Object.freeze({
    ...value,
    utxo: copyUtxo(value.utxo),
    outputContext: Object.freeze({
      ...value.outputContext,
      hash: copy(value.outputContext.hash),
    }),
    nullifier: copy(value.nullifier),
    ...(value.dataHash === undefined ? {} : { dataHash: copy(value.dataHash) }),
    ...(value.zoneDataHash === undefined ? {} : { zoneDataHash: copy(value.zoneDataHash) }),
  });
}

export class Wallet {
  readonly identity: ShieldedAddress;
  readonly #registry: AssetRegistry;
  #viewingKeyHistory: ViewingKeyEntry[];
  #utxos: WalletUtxo[] = [];
  #transactions: PrivateTransaction[] = [];
  #nullifiers = new Set<string>();
  #lastSynced = 0n;
  /**
   * Per-view-tag sync watermarks: for each tag, the indexer cursor up to which
   * every matching transaction has already been seen. Mirrors Rust's
   * `Wallet::sync_cursors`.
   *
   * Per tag rather than one shared position, because the tag set GROWS. Tags
   * come from a local counter plus a window, and a second device spending
   * beyond that window makes the counter lag by more than the window absorbs.
   * A tag learned late must be scanned from the beginning even though other
   * tags have advanced far past those slots, so a single cursor would skip
   * those transactions permanently.
   *
   * In memory only. A client persisting wallet state across restarts should
   * persist these too -- without them sync stays correct but pays for the full
   * history every time.
   */
  #syncCursors = new Map<string, Uint8Array>();
  /**
   * The same watermarks for the encrypted-utxo stream that proofless deposits
   * are read from. Mirrors Rust's `Wallet::proofless_cursors`.
   *
   * Separate from `#syncCursors` because they are positions in different
   * streams: reaching the tip of the transaction stream says nothing about
   * where the encrypted-utxo stream has been read to, and sharing one cursor
   * would skip rows in whichever stream is behind.
   */
  #prooflessCursors = new Map<string, Uint8Array>();
  /**
   * The same watermarks for the nullifier stream, keyed by nullifier rather than
   * by tag. Mirrors Rust's `Wallet::nullifier_cursors`.
   *
   * An entry here means the opposite of a tag cursor: a tag cursor records what
   * has been found, this records how far a spend has been confirmed *absent*.
   * Entries are dropped once the nullifier is spent, since the question is then
   * answered for good.
   */
  #nullifierCursors = new Map<string, Uint8Array>();

  constructor(input: Readonly<{ identity: ShieldedAddress; registry?: AssetRegistry }>) {
    this.identity = input.identity;
    this.#registry = input.registry?.clone() ?? new AssetRegistry();
    this.#viewingKeyHistory = [newViewingKeyEntry(input.identity.viewingPublicKey, 0n)];
  }

  get registry(): AssetRegistry {
    return this.#registry.clone();
  }

  get viewingKeyHistory(): readonly ViewingKeyEntry[] {
    return this.#viewingKeyHistory.map(snapshotViewingKeyEntry);
  }

  /** Timestamp the last completed sync was told to record, zero before the first. */
  get lastSynced(): bigint {
    return this.#lastSynced;
  }

  /**
   * Cursor this tag's stream has been read to, or `undefined` for a tag never
   * scanned -- which must start from the beginning, not from another tag's
   * position.
   *
   * @internal
   */
  _syncCursor(stream: CursorStream, tag: string): Uint8Array | undefined {
    return this.#cursorsFor(stream).get(tag);
  }

  /** @internal */
  _setSyncCursor(stream: CursorStream, tag: string, cursor: Uint8Array): void {
    this.#cursorsFor(stream).set(tag, Uint8Array.from(cursor));
  }

  /**
   * Forget the watermarks for nullifiers now known spent.
   *
   * A spent nullifier is never queried again, so its watermark is dead weight.
   * Without this the map grows with history even though the query set shrinks.
   *
   * @internal
   */
  _forgetNullifierCursors(spent: ReadonlySet<string>): void {
    for (const nullifier of spent) this.#nullifierCursors.delete(nullifier);
  }

  #cursorsFor(stream: CursorStream): Map<string, Uint8Array> {
    if (stream === "transactions") return this.#syncCursors;
    return stream === "proofless" ? this.#prooflessCursors : this.#nullifierCursors;
  }

  registerAsset(assetId: bigint, mint: Address): void {
    this.#registry.insert(assetId, mint);
  }

  utxos(): readonly WalletUtxo[] {
    return this.#utxos.map(snapshotUtxo);
  }

  privateTransactions(): readonly PrivateTransaction[] {
    return this.#transactions.map((transaction) =>
      Object.freeze({ ...transaction, id: Object.freeze({ ...transaction.id }) }),
    );
  }

  /**
   * The balance of one registered mint. A mint the wallet holds no note for
   * still has a balance of zero; only a mint the registry does not know is a
   * rejection.
   */
  balance(mint: Address, filter?: Filter): AssetBalance {
    const assetId = this.#registry.assetId(mint);
    const utxos = this.#utxos
      .filter(
        (entry) =>
          !entry.spent &&
          entry.utxo.asset === mint &&
          (filter === undefined || matches(filter, entry.utxo)),
      )
      .map((entry) => copyUtxo(entry.utxo));
    const amount = checkedBalance(utxos.reduce((sum, utxo) => sum + utxo.amount, 0n));
    return Object.freeze({ assetId, mint, amount, utxos: Object.freeze(utxos) });
  }

  /** One balance per mint the wallet holds an unspent note of, by asset id. */
  balances(skipUtxos = false): readonly AssetBalance[] {
    const mints = new Set(
      this.#utxos.filter((entry) => !entry.spent).map((entry) => entry.utxo.asset),
    );
    return [...mints]
      .map((mint) => {
        const balance = this.balance(mint);
        return skipUtxos ? Object.freeze({ ...balance, utxos: Object.freeze([]) }) : balance;
      })
      .sort((left, right) =>
        left.assetId < right.assetId ? -1 : left.assetId > right.assetId ? 1 : 0,
      );
  }

  /** @internal */
  _state(): Readonly<{
    utxos: readonly WalletUtxo[];
    transactions: readonly PrivateTransaction[];
    nullifiers: ReadonlySet<string>;
    viewingKeyHistory: readonly ViewingKeyEntry[];
  }> {
    return {
      utxos: this.#utxos.map(snapshotUtxo),
      transactions: this.privateTransactions(),
      nullifiers: new Set(this.#nullifiers),
      viewingKeyHistory: this.viewingKeyHistory,
    };
  }

  /**
   * Omitting `viewingKeyHistory` leaves retained viewing-key history untouched, and
   * omitting `lastSynced` leaves the sync timestamp untouched.
   * @internal
   */
  _replace(
    input: Readonly<{
      utxos: readonly WalletUtxo[];
      transactions: readonly PrivateTransaction[];
      nullifiers: ReadonlySet<string>;
      viewingKeyHistory?: readonly ViewingKeyEntry[];
      lastSynced?: bigint;
    }>,
  ): void {
    const hashes = new Set<string>();
    for (const entry of input.utxos) {
      const hash = hex(entry.outputContext.hash);
      if (hashes.has(hash)) {
        throw new TransactionError("TRANSACTION_DUPLICATE_OUTPUT", { hash });
      }
      hashes.add(hash);
    }
    this.#utxos = input.utxos.map(snapshotUtxo);
    this.#transactions = input.transactions.map((transaction) =>
      Object.freeze({ ...transaction, id: Object.freeze({ ...transaction.id }) }),
    );
    this.#nullifiers = new Set(input.nullifiers);
    // A spent nullifier is never queried again, so its watermark is dead weight.
    // Both key spaces are lowercase hex of the same bytes, so the sets line up.
    this._forgetNullifierCursors(this.#nullifiers);
    if (input.viewingKeyHistory !== undefined) {
      this.#viewingKeyHistory = input.viewingKeyHistory.map(snapshotViewingKeyEntry);
    }
    if (input.lastSynced !== undefined) {
      this.#lastSynced = input.lastSynced;
    }
  }
}

export function hex(bytes: Uint8Array): string {
  return [...bytes].map((byte) => byte.toString(16).padStart(2, "0")).join("");
}

function checkedBalance(amount: bigint): bigint {
  if (amount > 0xffff_ffff_ffff_ffffn) {
    throw new TransactionError("TRANSACTION_WALLET_BALANCE_OVERFLOW");
  }
  return amount;
}
